docs/designs/ps_design.md

```diff
@@ -1,375 +1,191 @@
 
-![pserver](../images/pserver.png)
+We use a KVStore to store model parameters in PS. There is a KVStore instance in each PS pod. The KVStore instances from the PS pods forms a distributed KVStore, which could be scaled easily to support a model with a large size.
 
-PS contains two main compoments:
+We also need an optimizer to update the model parameters stored in PS pods. To update a single parameter, the optimizer needs to get the parameter, apply the gradient to the parameter, and then write the parameter back. It involves one time read and one time write. There will be huge accesses to parameters during a training job. Since model parameters are store at PS, it's better to make the optimization at the same place to reduce the cost of accessing parameters.
 
-- KVStore
-- Optimizer
+Besides KVStore and optimizer, the PS also need to provide necessary RPC services to workers. Workers will pull the latest model parameters from PS, and push gradients to PS in each iteration of training.
 
-The worker initializes a model, and pushes parameters to KVStore. Before each step of training, the worker pulls the latest model from KVStore. After a round of forward/backward computation, the worker pushes gradients to the PS waiting for processing. Then, optimizer of PS will look up the corresponding parameter from KVStore. At last, it applies gradients to parameters, and updates parameter back to KVStore.
+Thus, we propose that PS has three basic components:
 
+| component| functionality |hardware |
+| :----  |:----  |:----  |
+|KVStore  | provides distributed storage of model parameters|memory|
+|Optimizer | applies gradients from workers to model parameters|CPU |
+|RPC servicer |servers workers to pull parameters and push gradients | network bandwidth |
 
-### KVStore
+Following is the architecture diagram of PS:
 
-The KVStore could be like following:
+![pserver](../images/pserver.png)
 
-```python
-class KVStore(object):
-    def __init__(self):
-        self.variable_db = {}
-        self.embedding_table_db = {}
 
-    def get_parameter(self, name):
-        pass
+### KVStore
 
-    def get_embedding_vector(self, name, indices):
-        pass
+The KVStore needs to support both embedding vector parameter and non-embedding parameter.
 
-    def set_parameter(self, name, value):
-        pass
+Since `tf.keras.optimizer` only accept `tf.Variable` as the parameter. To avoid unnecessary memory copy, we save non-embedding parameter as a `tf.Variable` in variable DB directly.
 
-    def set_embedding_vector(self, name, indices, value):
-        pass
-```
+For the embedding vector parameter, there are two reasons that we could not initialize it before training:
+
+- Feature ID is usually calculated by the hash function. The training dataset is too large to get the range of the ID before training.
+- Sometimes, there comes a new feature ID in online learning.
```

Comment:
You mean the set of feature IDs cannot be defined at once and is growing?

Reply:
Yes, so the size embedding table of feature ID could not be decided. We could not take it as a common `tf.Variable`, we need a customized data structure to handle it.